sway-error/src/error.rs

```diff
@@ -3229,6 +3246,49 @@ impl ToDiagnostic for CompileError {
                 hints: vec![],
                 help: vec![format!("The name must be a valid Sway identifier{}.", if name.is_empty() { " and cannot be empty" } else { "" })],
             },
+            IndexedFieldInNonEventStruct { field_name, struct_name } => Diagnostic {
+                reason: Some(Reason::new(code(1), "Indexed fields must be in event structs".to_string())),
+                issue: Issue::error(
+                    source_engine,
+                    field_name.span(),
+                    format!("Field \"{field_name}\" is marked as `#[indexed]`, but its struct is not an event.")
+                ),
+                hints: vec![
+                    Hint::help(
+                        source_engine,
+                        struct_name.span(),
+                        format!("Consider annotating \"{struct_name}\" struct with the `#[event]` attribute."),
+                    )
+                ],
+                help: vec![
+                    "Fields can be marked as `#[indexed]` only if their parent struct is annotated with the `#[event]` attribute.".to_string(),
+                ]
+            },
+            IndexedFieldMustPrecedeNonIndexedField { field_name } => Diagnostic {
+                reason: Some(Reason::new(code(1), "Indexed fields must precede non-indexed fields".to_string())),
+                issue: Issue::error(
+                    source_engine,
+                    field_name.span(),
+                    format!("Field \"{field_name}\" cannot be marked as `#[indexed]` because a non-indexed field appears before it."),
+                ),
+                hints: vec![],
+                help: vec![
+                    "Event structs must list their `#[indexed]` fields first, followed by non-indexed fields.".to_string(),
+                ],
+            },
+            IndexedFieldOffsetTooLarge { field_name } => Diagnostic {
```

Comment:
nit: should we add a `to_diagnostic` impl match for `IndexedFieldIsNotFixedSizeABIType`? Otherwise it will just hit the catch-all; maybe something like so?
```suggestion
            IndexedFieldIsNotFixedSizeABIType { field_name } => Diagnostic {
                reason: Some(Reason::new(code(1), "Indexed fields must be fixed-size ABI types".to_string())),
                issue: Issue::error(
                    source_engine,
                    field_name.span(),
                    format!("Field \"{field_name}\" is marked as `#[indexed]`, but is not a fixed-size ABI type."),
                ),
                hints: vec![],
                help: vec![
                    "Indexed fields in event structs must be fixed-size ABI types.".to_string(),
                    "Fixed-size types include: bool, u8, u16, u32, u64, u256, b256, and addresses.".to_string(),
                ],
            },
            IndexedFieldOffsetTooLarge { field_name } => Diagnostic {
```